lang/parser/src/cst/decls.rs

```diff
@@ -30,104 +31,181 @@ pub enum Decl {
 pub struct Ctor {
     pub span: Span,
     pub doc: Option<DocComment>,
-    pub name: exp::Ident,
+    pub name: Ident,
     pub params: Telescope,
     pub typ: Option<TypApp>,
 }
 
+/// Declaration of a destructor within the context of a codata type declaration.
+///
+/// ```text
+/// codata F(...) { (self: F(...)).d(...) : t }
+///                 ^^^^^^^^^^^^^^ ^  ^     ^
+///                       |        |  |     \----- ret_typ
+///                       |        |  \----------- params
+///                       |        \-------------- name
+///                       \----------------------- destructee
+/// ```
 #[derive(Debug, Clone)]
 pub struct Dtor {
     pub span: Span,
     pub doc: Option<DocComment>,
-    pub name: exp::Ident,
+    pub name: Ident,
     pub params: Telescope,
     pub destructee: Destructee,
     pub ret_typ: Rc<exp::Exp>,
 }
 
+/// Destructee within the context of a destructor declaration in a codata type.
+///
+/// ```text
+/// codata F(...) { (self: F(...)).d(...) : t }
+///                  ^^^^  ^^^^^^
+///                    |      \----- typ
+///                    \------------ name
+/// ```
+#[derive(Debug, Clone)]
+pub struct Destructee {
+    pub span: Span,
+    pub name: Option<Ident>,
+    pub typ: Option<TypApp>,
+}
+
+/// Toplevel definition, i.e. a global pattern match.
+///
+/// ```text
+/// def (self: F(...)).d(...) : t { ... }
+///     ^^^^^^^^^^^^^^ ^  ^     ^    ^
+///            |       |  |     |    \----- body
+///            |       |  |     \---------- ret_typ
+///            |       |  \---------------- params
+///            |       \------------------- name
+///            \--------------------------- scrutinee
+/// ```
 #[derive(Debug, Clone)]
 pub struct Def {
     pub span: Span,
     pub doc: Option<DocComment>,
-    pub name: exp::Ident,
+    pub name: Ident,
     pub attr: Attribute,
     pub params: Telescope,
     pub scrutinee: Scrutinee,
     pub ret_typ: Rc<exp::Exp>,
     pub body: exp::Match,
 }
 
+/// Scrutinee within a toplevel definition
+///
+/// ```text
+/// def (self: F(...)).d(...) : t { ... }
+///      ^^^^  ^^^^^^
+///        |      \----- typ
+///        \------------ name
+/// ```
+#[derive(Debug, Clone)]
+pub struct Scrutinee {
+    pub span: Span,
+    pub name: Option<Ident>,
+    pub typ: TypApp,
+}
+
+/// Toplevel codefinition, i.e. a global copattern match.
+///
+/// ```text
+/// codef C(...) : F(...) { ... }
+///       ^  ^     ^^^^^^    ^
+///       |  |        |      \------ body
+///       |  |        \------------- typ
+///       |  \---------------------- params
+///       \------------------------- name
+/// ```
 #[derive(Debug, Clone)]
 pub struct Codef {
     pub span: Span,
     pub doc: Option<DocComment>,
-    pub name: exp::Ident,
+    pub name: Ident,
     pub attr: Attribute,
     pub params: Telescope,
     pub typ: TypApp,
     pub body: exp::Match,
 }
 
+/// Toplevel let-bound expression.
+///
+/// ```text
+/// let I(...) : t { e }
```

Comment:
Nitpick, but shouldn't top-level let expressions be lower-case? One can view them as syntax sugar for top-level pattern-matching definitions on `Top`, meaning they should be lowercase.

Example:

```
let foo(...) : t { e }
```

can also be thought of as:

```
def Top.foo(...): t {
    Unit => e
}
```

Reply:
I was thinking about this, but didn't come to a good conclusion: Should we be able to distinguish variables from calls directly in the parser or not? We certainly know that when we see `f(a,b,c)`, then `f` cannot be a bound variable. But what about just `f`?  I see multiple alternatives:
- Just postpone this to the lowering phase. This is the most permissive solution, but complicates lowering since we have to look up both in the local context and in the global context, and locallly bound variables can shadow global definitions.
- Without parentheses `f` is always a local variable, if we want to refer to stuff from the global context then we require empty parens: `f()`. But I certainly don't want to write `True()` instead of `True`!
- Force all toplevel producers (i.e. constructors, let-bound definitions and codefinitions) to be uppercase, and variables to be lowercase. Then `f` must always be a variable, `f(...)` on its own is not valid and only `e.f(...)` is allowed.

Reply:
I strongly advocate for dealing with this in lowering. This is exactly the kind of thing lowering is for IMHO.

Reply:
I changed to a lowercase letter in [791ee57](https://github.com/polarity-lang/polarity/pull/202/commits/791ee57af475d01fc1e19bf861f30081b55f576e)